src/Function/Bundles/Related.agda

```diff
@@ -0,0 +1,399 @@
+------------------------------------------------------------------------
+
+-- The Agda standard library
+--
+-- Relatedness for the function hierarchy
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Function.Bundles.Related where
+
+open import Level
+open import Relation.Binary using (Sym; Reflexive; Trans; IsEquivalence; Setoid; IsPreorder; Preorder)
+open import Function.Bundles
+open import Function.Base
+open import Function.Equality using (_⟨$⟩_)
+import Function.Equivalence as Eq
+import Function.Injection   as Inj
+import Function.Inverse     as Inv
+import Function.LeftInverse as LeftInv
+import Function.Surjection  as Surj
+import Relation.Binary.PropositionalEquality as P
+open import Function.Properties.Bijection
+open import Function.Properties.Surjection
+open import Function.Properties.Inverse
+open import Function.Properties.RightInverse
+
+import Function.Construct.Symmetry as Symmetry
+import Function.Construct.Identity as Identity
+import Function.Construct.Composition as Composition
+
+open import Data.Product
+
+------------------------------------------------------------------------
+-- Relatedness
+
+-- There are several kinds of "relatedness".
+
+-- The idea to include kinds other than equivalence and bijection came
+-- from Simon Thompson and Bengt Nordström. /NAD
+
+data Kind : Set where
+  implication         : Kind
+  reverse-implication : Kind
+  equivalence         : Kind
+  injection           : Kind
+  reverse-injection   : Kind
+  left-inverse        : Kind
+  surjection          : Kind
+  bijection           : Kind
+
+-- Interpretation of the codes above. The code "bijection" is
+-- interpreted as Inverse rather than Bijection; the two types are
+-- equivalent.
+
+infix 4 _∼[_]_
+
+_∼[_]_ : ∀ {ℓ₁ ℓ₂} → Set ℓ₁ → Kind → Set ℓ₂ → Set _
+A ∼[ implication         ] B = A ⟶ B
+A ∼[ reverse-implication ] B = B ⟶ A
+A ∼[ equivalence         ] B = A ⇔ B
+A ∼[ injection           ] B = A ↣ B
+A ∼[ reverse-injection   ] B = B ↣ A
+A ∼[ left-inverse        ] B = A ↪ B
+A ∼[ surjection          ] B = A ↠ B
+A ∼[ bijection           ] B = A ⤖ B
+
+-- A non-infix synonym.
+
+Related : Kind → ∀ {ℓ₁ ℓ₂} → Set ℓ₁ → Set ℓ₂ → Set _
+Related k A B = A ∼[ k ] B
+
+-- The bijective equality implies any kind of relatedness.
+
+⤖⇒ : ∀ {k x y} {X : Set x} {Y : Set y} →
```

Comment:
We should stick with `A` and `B` to be consistent rather than switching to `X` and `Y`.